result.nim

```diff
@@ -0,0 +1,352 @@
+    else:
+      new (e)
+    raise e
+  self.value
+
+proc `[]`*[T](self: Result[T, ref Exception]): T =
+  ## Fetch value of result if set, or raise the contained exception
+  if self.isErr:
+    raise self.error
+  self.value
+
+proc `$`*[T, E](self: Result[T, E]): string =
+  ## Returns string representation of `self`
+  if self.isOk: "Ok(" & $self.value & ")"
+  else: "Err(" & $self.error & ")"
+
+template derefOr[T, E](self: Result[T, E], def: T): T =
+  ## Fetch value of result if set, or supplied default
+  ## default will not be evaluated iff value is set
+  if self.isErr: def
+  else: self.value
+
+iterator items[T, E](self: Result[T, E]): T =
+  ## Iterate over result as if it were a collection of either 0 or 1 items
+  if self.isOk:
+    yield self.value
+
+when isMainModule:
+  type R = Result[int]
+
+  # Basic usage, producer
+  proc works(): R =
+    R.ok(42)
+  proc fails(): R =
+    R.err("dummy")
+
+  proc works2(): R =
+    result.ok(42)
+  proc fails2(): R =
+    result.err("dummy")
+
+  proc raises(): int =
+    raise newException(Exception, "hello")
+
+  # Basic usage, consumer
+  let
+    a = works()
+    b = fails()
+
+  doAssert a.isOk
+  doAssert a.value == 42
+  doAssert (not a.isErr)
+
+  # Combine
+  let xxx = a and b
+  doAssert (not xxx.isOk)
+  doAssert (a or b).isOk
+
+  # Exception on access
+  let va = try: discard a.error; false except: true
+  doAssert va, "not an error, should raise"
+
+  # Exception on access
+  let vb = try: discard b.value; false except: true
+  doAssert vb, "not an error, should raise"
+
+  var x = a
+
+  # Mutate
+  x.err("failed now")
+
+  doAssert x.isErr
+
+  # Exceptions -> results
+  let c = catch:
+    raises()
+
+  doAssert c.isErr
+
+  # De-reference
+  try:
+    echo b[]
+    doAssert false
+  except:
+    discard
+
+  doAssert a.derefOr(50) == a.value
+  doAssert b.derefOr(50) == 50
+
+  # Comparisons
+  doAssert (works() == works2())
+  doAssert (fails() == fails2())
+  doAssert (works() != fails())
+
+  # Simple lifting..
+  doAssert (a + a)[] == a.value + a.value
+
+  var counter = 0
+  proc incCounter(): R =
+    counter += 1
+    R.ok(counter)
+
+  doAssert (b and incCounter()).isErr, "b fails"
+  doAssert counter == 0, "should fail fast on b"
+
+  # Iteration
+  var counter2 = 0
+  for v in a:
+    counter2 += 1
+
+  doAssert counter2 == 1, "one-item collection when set"
+
+  # Mapping
+  doAssert (a.map(proc(x: int): string = $x)[] == $a.value)
+  doAssert (b.mapErr(proc(x: string): string = x & "no!").error == (b.error & "no!"))
+
+  # Exception interop
+  let e = capture(int, newException(Exception, "test"))
+  doAssert e.isErr
+  try:
+    discard e[]
+    doAssert false, "should have raised"
+  except:
+    doAssert getCurrentException().msg == "test"
+
+  # Nice way to checks
+  if (let v = works(); v.isOk):
+    doAssert v[] == v.value
+
+  doAssert $a == "Ok(42)"
+
+  doAssert a.mapConvert(int64)[] == int64(42)
+  doAssert a.mapCast(int8)[] == int8(42)
```

Comment:
All in all, looks pretty great. 

For the intended support in `handleErrors` I imagined one more possible feature. There is a planned support in Nim for `varargs` generic types where the `varargs` parameter is mapped to a tuple (it's a relatively simple feature). We can then possibly add the option to instantiate `Result` with multiple possible error types (one of the goals of `handleErrors` is to make it easy to switch between `enum` errors and richer error types that can hold values).